Watchman.Discord/Areas/Users/Services/WarnsService.cs

```diff
@@ -0,0 +1,83 @@
+﻿using Devscord.DiscordFramework.Commons.Exceptions;
+using Devscord.DiscordFramework.Commons.Extensions;
+using Devscord.DiscordFramework.Framework.Commands.Responses;
+using Devscord.DiscordFramework.Middlewares.Contexts;
+using Devscord.DiscordFramework.Services;
+using Devscord.DiscordFramework.Services.Factories;
+using MongoDB.Driver;
+using MongoDB.Driver.Core.Servers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Watchman.Cqrs;
+using Watchman.Discord.Areas.Users.BotCommands;
+using Watchman.Discord.Areas.Users.BotCommands.Warns;
+using Watchman.DomainModel.Messages;
+using Watchman.DomainModel.Warns;
+using Watchman.DomainModel.Warns.Commands;
+using Watchman.DomainModel.Warns.Queries;
+
+namespace Watchman.Discord.Areas.Protection.Services
+{
+    public class WarnsService
+    {
+        private readonly UsersService _usersService;
+        private readonly MessagesServiceFactory _messagesServiceFactory;
+        private readonly DirectMessagesService _directMessagesService;
+        private readonly ICommandBus _commandBus;
+        private readonly IQueryBus _queryBus;
+
+        public WarnsService(UsersService usersService, MessagesServiceFactory messagesServiceFactory, DirectMessagesService directMessagesService, ICommandBus commandBus, IQueryBus queryBus)
+        {
+            this._messagesServiceFactory = messagesServiceFactory;
+            this._directMessagesService = directMessagesService;
+            this._usersService = usersService;
+            this._commandBus = commandBus;
+            this._queryBus = queryBus;
+        }
+
+        public Task AddWarnToUser(AddWarnCommand command, Contexts contexts, UserContext targetUser)
+        {
+            var addWarnEventCommand = new AddWarnEventCommand(contexts.User.Id, targetUser.Id, command.Reason, contexts.Server.Id);
+            return this._commandBus.ExecuteAsync(addWarnEventCommand);
+        }
+
+        public IEnumerable<KeyValuePair<string, string>> GetWarns(WarnsCommand command, Contexts contexts, UserContext mentionedUser, ulong serverId)
+        {
+            if (mentionedUser == null)
+            {
+                return null;
+            }
+            var warnEvents = GetWarnEvents(serverId, mentionedUser.Id).GetAwaiter().GetResult();
+            return WarnEventsToKeyValue(warnEvents, mentionedUser.Id);
+        }
+
+        public async Task<IEnumerable<WarnEvent>> GetWarnEvents(ulong serverId, ulong userId)
+        {
+            var query = new GetWarnEventsQuery(serverId, userId);
+            var response = await this._queryBus.ExecuteAsync(query);
```

Comment:
nie lepiej użyć synchronicznej wersji Execute(query)?

Reply:
obojętnie, pytanie ile będzie trwać zapytanie o wszystkie ostrzeżenia dla kogoś

Reply:
i tak musi czekać na te informacje, więc dużo to nie zmienia

Reply:
nie jestem pewien ale chyba jeśli jest async, to inny user może w tym czasie wpisać kolejną komendę

Reply:
i tak może, bo każda komenda to oddzielne wywołanie przez delegat
poza tym pobranie tego z bazy to kilkadziesiąt ms (w skrajnych przypadkach kilkaset ms)